util/linuxfw/nftables_runner.go

```diff
@@ -0,0 +1,1068 @@
+				Register: 1,
+			},
+			&expr.Cmp{
+				Op:       expr.CmpOpEq,
+				Register: 1,
+				Data:     []byte("lo"),
+			},
+			saddrExpr,
+			&expr.Cmp{
+				Op:       expr.CmpOpEq,
+				Register: 1,
+				Data:     addr.AsSlice(),
+			},
+			&expr.Counter{},
+			&expr.Verdict{
+				Kind: expr.VerdictAccept,
+			},
+		},
+	}
+
+	existingLoopBackRule, err := findRule(conn, loopBackRule)
+	if err != nil {
+		return fmt.Errorf("find loop back rule: %w", err)
+	}
+
+	if err := conn.DelRule(existingLoopBackRule); err != nil {
+		return fmt.Errorf("delete rule: %w", err)
+	}
+
+	if err := conn.Flush(); err != nil {
+		return fmt.Errorf("flush: %w", err)
+	}
+
+	return nil
+}
+
+// getTables gets the available nftable in nftables runner.
+func (n *nftablesRunner) getTables() []*nftable {
+	if n.v6Available {
+		return []*nftable{n.nft4, n.nft6}
+	}
+	return []*nftable{n.nft4}
+}
+
+// getNATTables gets the available nftable in nftables runner.
+// If the system does not support IPv6 NAT, only the IPv4 nftable
+// will be returned.
+func (n *nftablesRunner) getNATTables() []*nftable {
+	if n.v6NATAvailable {
+		return n.getTables()
+	}
+	return []*nftable{n.nft4}
+}
+
+// AddChains creates custom Tailscale chains in netfilter via nftables
+// if the ts-chain doesn't already exist.
+func (n *nftablesRunner) AddChains() error {
+	conn := n.conn
+
+	for _, table := range n.getTables() {
+		filter, err := createTableIfNotExist(conn, table.Proto, "ts-filter")
+		if err != nil {
+			return fmt.Errorf("create table: %w", err)
+		}
+		table.Filter = filter
+		if err = createChainIfNotExist(conn, filter, tsChain("forward"), nftables.ChainTypeFilter, nftables.ChainHookForward, nftables.ChainPriorityRef(-1)); err != nil {
+			return fmt.Errorf("create forward chain: %w", err)
+		}
+		if err = createChainIfNotExist(conn, filter, tsChain("input"), nftables.ChainTypeFilter, nftables.ChainHookInput, nftables.ChainPriorityRef(-1)); err != nil {
+			return fmt.Errorf("create input chain: %w", err)
+		}
+	}
+
+	for _, table := range n.getNATTables() {
+		nat, err := createTableIfNotExist(conn, table.Proto, "ts-nat")
+		if err != nil {
+			return fmt.Errorf("create table: %w", err)
+		}
+		table.Nat = nat
+		if err = createChainIfNotExist(conn, nat, tsChain("postrouting"), nftables.ChainTypeNAT, nftables.ChainHookPostrouting, nftables.ChainPriorityNATDest); err != nil {
+			return fmt.Errorf("create postrouting chain: %w", err)
+		}
+	}
+
+	if err := conn.Flush(); err != nil {
+		return fmt.Errorf("flush: %w", err)
+	}
+
+	return nil
+}
+
+// deleteChainIfExists deletes a chain if it exists.
+func deleteChainIfExists(c *nftables.Conn, table *nftables.Table, name string) error {
+	chain, err := getChainFromTable(c, table, name)
+	if err != nil {
+		return fmt.Errorf("get chain: %w", err)
+	}
+
+	c.FlushChain(chain)
+	c.DelChain(chain)
+
+	if err := c.Flush(); err != nil {
+		return fmt.Errorf("flush and delete chain: %w", err)
+	}
+
+	return nil
```

Comment:
I think the error message here could be a little more clear - this is flushing but not deleting (or just do a return c.Flush())